plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts

```diff
@@ -321,6 +478,304 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
+        email: event.email,
+        name: event.name,
+        state: 'active', // in the delete case it doesn't really matter if the user is active or not
+        web_url: '',
+        avatar_url: '',
+      };
+    } else {
+      const client = new GitLabClient({
+        config: this.integration.config,
+        logger: this.logger,
+      });
+
+      user = await client.getUserById(event.user_id);
+    }
+
+    if (!this.shouldProcessUser(user)) {
+      this.logger.debug(`Skipped user ${user.username}.`);
+      return;
+    }
+
+    const userEntities: UserEntityV1alpha1[] = [];
+    userEntities.push(
+      await this.userTransformer({
+        user: user,
+        integrationConfig: this.integration.config,
+        providerConfig: this.config,
+        groupNameTransformer: this.groupNameTransformer,
+      }),
+    );
+
+    const { added, removed } = createDeltaOperation([...userEntities]);
+
+    this.logger.debug(`Applying mutation for user ${user.username}.`);
+    await this.connection.applyMutation({
+      type: 'delta',
+      removed,
+      added,
+    });
+  }
+
+  // the goal here is to reconstruct the group either from which the user was removed or to which the user was added. Specifically, we add/remove the new user to/from the spec.member property array of the group entity. The Processor should take care of updating the relations
+  private async onMembershipChange(
+    event: SystemHookCreateOrDestroyMembershipEventsSchema,
+    createDeltaOperation: (entities: Entity[]) => {
+      added: any[];
+      removed: any[];
+    },
+  ): Promise<void> {
+    if (!this.connection) {
+      throw new Error(
+        `Gitlab discovery connection not initialized for ${this.getProviderName()}`,
+      );
+    }
+
+    const client = new GitLabClient({
+      config: this.integration.config,
+      logger: this.logger,
+    });
+
+    const groupToRebuild: GitLabGroup = await client.getGroupById(
+      event.group_id,
+    );
+
+    // If the group is outside the scope there is no point creating anything related to it.
+    if (!this.shouldProcessGroup(groupToRebuild)) {
+      this.logger.debug(`Skipped group ${groupToRebuild.full_path}.`);
+      return;
+    }
+
+    // get members of the group
+    const usersToBeAdded: string[] = [];
+    let groupMembers: PagedResponse<GitLabUser> = { items: [] };
+
+    const relations: string[] =
+      this.config.allowInherited === true && this.config.allowInherited
+        ? ['DIRECT', 'INHERITED']
+        : ['DIRECT'];
+
+    groupMembers = await client.getGroupMembers(
+      groupToRebuild.full_path,
+      relations,
+    );
+
+    if (groupMembers.items.length !== 0) {
+      groupMembers.items.forEach(element => {
+        if (
+          event.event_name === 'user_remove_from_group' &&
+          element.username === event.user_username
+        ) {
+          return;
+        }
+
+        usersToBeAdded.push(element.username);
+      });
+    }
```

Comment:
if I've read the logic correctly this could be simplified a bit:

```suggestion
    const relations = this.config.allowInherited
      ? ['DIRECT', 'INHERITED']
      : ['DIRECT'];

    const groupMembers = await client.getGroupMembers(
      groupToRebuild.full_path,
      relations,
    );

    // new members of the group
    const usersToBeAdded =
      event.event_name === 'user_remove_from_group'
        ? groupMembers.items.filter(e => e.username !== event.user_username)
        : groupMembers.items;
    }
```

then below
```js
usersToBeAdded.map(e => e.username)
```

Reply:
Great remark, thanks! :)